File: Assets/Scripts/FlickeringLight.lua
local CS = CS
local UnityEngine = CS.UnityEngine
local Time = UnityEngine.Time

local minIntensity = 0.5
local maxIntensity = 7.0
local minInterval = 9.0
local maxInterval = 15.0

local flickeringLight = nil
local audioSource = nil
local nextFlickerTime = 0.0
local flickerEndTime = 0.0
local isFlickering = false
local flickerDuration = 0.7

function calculateValue(base, range, factor)
    return base + range * math.abs(math.sin(Time.time * factor))
end

function awake()
    flickeringLight = gameObject:GetComponent(typeof(UnityEngine.Light))
    audioSource = gameObject:GetComponent(typeof(UnityEngine.AudioSource))

    if flickeringLight then
        flickeringLight.intensity = maxIntensity
    end

    nextFlickerTime = Time.time + calculateValue(minInterval, maxInterval - minInterval, 0.1)
end

function update()
    local currentTime = Time.time

    if not isFlickering and currentTime >= nextFlickerTime then
        isFlickering = true
        flickerDuration = calculateValue(0.1, 0.9, 0.2)
        flickerEndTime = currentTime + flickerDuration
    end

    if isFlickering then
        if currentTime < flickerEndTime then
            if flickeringLight then
                local flickerFactor = math.random(5, 15) * 0.1
                flickeringLight.intensity = calculateValue(minIntensity, maxIntensity - minIntensity, flickerFactor)
            end
            if audioSource and flickerSound then
                audioSource:PlayOneShot(flickerSound)
            end
        else
            if audioSource and audioSource.isPlaying then
                audioSource:Stop()
            end
            isFlickering = false
            if flickeringLight then
                flickeringLight.intensity = maxIntensity
            end
            nextFlickerTime = currentTime + calculateValue(minInterval, maxInterval - minInterval, 0.1)
        end
    end
end

File: Assets/Scripts/Heights Scene/FlickeringLight.cs
using UnityEngine;

public class FlickeringLight : MonoBehaviour
{
    [SerializeField] private float minIntensity = 0.5f;
    [SerializeField] private float maxIntensity = 1.5f;
    [SerializeField] private float flickerSpeed = 0.1f;
    [SerializeField] private float minInterval = 4f;
    [SerializeField] private float maxInterval = 8f;
    [SerializeField] private AudioClip flickerSound;

    private float nextFlickerTime = 0f;
    private float flickerEndTime = 0f;
    private bool isFlickering = false;
    private Light flickeringLight;
    private float flickerDuration = 0.7f;
    private AudioSource audioSource;

    void Start()
    {
        flickeringLight = GetComponent<Light>();
        flickeringLight.intensity = maxIntensity;
        nextFlickerTime = Time.time + Random.Range(minInterval, maxInterval);
        audioSource = GetComponent<AudioSource>();
    }

    private void Update()
    {
        float currentTime = Time.time;

        if (!isFlickering && currentTime >= nextFlickerTime)
        {
            isFlickering = true;
            flickerDuration = Random.Range(0.1f, 1.0f);
            flickerEndTime = currentTime + flickerDuration;
        }

        if (isFlickering)
        {
            if (currentTime < flickerEndTime)
            {
                flickeringLight.intensity = Random.Range(minIntensity, maxIntensity);
                audioSource.PlayOneShot(flickerSound);
            }
            else
            {
                if(audioSource.isPlaying)
                    audioSource.Stop();
                isFlickering = false;
                flickeringLight.intensity = maxIntensity;
                nextFlickerTime = currentTime + Random.Range(minInterval, maxInterval);
            }
        }
    }
}